feat: `gnome-extensions` module

fiftydinar opened the pull request:
If you wished to install some Gnome extension, but couldn't find it in RPM, this extension will help you in that.

Basically all extensions are supported, with some rare extensions like Pano just additionally needing some system dependencies.

See README for more details.

PR is in draft as it just needs some cosmetic refinements in some areas & probably some documentation fixes.

It's not a big code, so I believe that it will suit the future bash module code guidelines nicely.

Some check like extension compatibility with Gnome version can also be implemented to fail if detected, based on metadata.json. I forgot that I could implement that.

EDIT: Implemented the compatibility check between extension & Gnome version.

fiftydinar committed:
feat: `gnome-extensions` module

If you wished to install some Gnome extension, but couldn't find it in RPM, this extension will help you in that.

Basically all extensions are supported, with some rare extensions like Pano just additionally needing some system dependencies.

See README for more details.

xynydev commented on modules/gnome-extensions/README.md:
```diff
@@ -0,0 +1,24 @@
+This module is universally compatible with all distributions which ship Gnome, as long as some extension doesn't require some additional dependency, like Pano.
+
+Thanks to https://extensions.gnome.org which provides end-releases of extensions as zips, it is very easy to maintain this module configuration.
+Basically the only maintenance is to bump the extension version when new Fedora/Gnome releases (around every 6 months).
```
This sparked me to do some inspecting and debugger digging on the extensions page. I was able to find that an extension page on https://extensions.gnome.org/ contains a div with the data of all the versions, for example: 
```html
<div class="extension single-page dummy" data-epk="6" data-uuid="apps-menu@gnome-shell-extensions.gcampax.github.com" data-svm="{'3.2': {'pk': 284, 'version': 5}, '3.2.0': {'pk': 6, 'version': 1}, '3.3.2': {'pk': 33, 'version': 3}, '3.2.1': {'pk': 52, 'version': 4}, '3.2.3': {'pk': 284, 'version': 5}, '3.3.4': {'pk': 640, 'version': 7}, '3.3.90': {'pk': 831, 'version': 8}, '3.4': {'pk': 1008, 'version': 9}, '3.5.5': {'pk': 1690, 'version': 10}, '3.5.90': {'pk': 1736, 'version': 11}, '3.5.91': {'pk': 1807, 'version': 12}, '3.6': {'pk': 2213, 'version': 14}, '3.7.5': {'pk': 2489, 'version': 16}, '3.7.90': {'pk': 2560, 'version': 17}, '3.7.92': {'pk': 2640, 'version': 18}, '3.8': {'pk': 3216, 'version': 21}, '3.10': {'pk': 3247, 'version': 22}, '3.12': {'pk': 3739, 'version': 23}, '3.14': {'pk': 4332, 'version': 24}, '3.15.1': {'pk': 4463, 'version': 25}, '3.15.3': {'pk': 4582, 'version': 26}, '3.16': {'pk': 4763, 'version': 27}, '3.18': {'pk': 5320, 'version': 29}, '3.20': {'pk': 5764, 'version': 30}, '3.21.2': {'pk': 5980, 'version': 31}, '3.22': {'pk': 6209, 'version': 32}, '3.24': {'pk': 6867, 'version': 34}, '3.26': {'pk': 7470, 'version': 36}, '3.28': {'pk': 8121, 'version': 38}, '3.30': {'pk': 8381, 'version': 39}, '3.32': {'pk': 10201, 'version': 41}, '3.34': {'pk': 13363, 'version': 43}, '3.36': {'pk': 14372, 'version': 44}, '3.38': {'pk': 21648, 'version': 46}, '40.0': {'pk': 22834, 'version': 47}, '40': {'pk': 24975, 'version': 48}, '41': {'pk': 28466, 'version': 50}, '42': {'pk': 29906, 'version': 51}, '43': {'pk': 34582, 'version': 52}, '44': {'pk': 40512, 'version': 53}, '45': {'pk': 48949, 'version': 56}, '46': {'pk': 52131, 'version': 58}}" data-versions="{'3.2': {'6': {'pk': 6, 'version': '1'}, '52': {'pk': 52, 'version': '4'}, '284': {'pk': 284, 'version': '5'}}, '3.2.0': {'6': {'pk': 6, 'version': '1'}}, '3.3.2': {'33': {'pk': 33, 'version': '3'}}, '3.2.1': {'52': {'pk': 52, 'version': '4'}}, '3.2.3': {'284': {'pk': 284, 'version': '5'}}, '3.3.4': {'640': {'pk': 640, 'version': '7'}}, '3.3.90': {'831': {'pk': 831, 'version': '8'}}, '3.4': {'1008': {'pk': 1008, 'version': '9'}}, '3.5.5': {'1690': {'pk': 1690, 'version': '10'}}, '3.5.90': {'1736': {'pk': 1736, 'version': '11'}}, '3.5.91': {'1807': {'pk': 1807, 'version': '12'}}, '3.6': {'1972': {'pk': 1972, 'version': '13'}, '2213': {'pk': 2213, 'version': '14'}}, '3.7.5': {'2473': {'pk': 2473, 'version': '15'}, '2489': {'pk': 2489, 'version': '16'}}, '3.7.90': {'2560': {'pk': 2560, 'version': '17'}}, '3.7.92': {'2640': {'pk': 2640, 'version': '18'}}, '3.8': {'2691': {'pk': 2691, 'version': '19'}, '2971': {'pk': 2971, 'version': '20'}, '3216': {'pk': 3216, 'version': '21'}}, '3.10': {'3247': {'pk': 3247, 'version': '22'}}, '3.12': {'3739': {'pk': 3739, 'version': '23'}}, '3.14': {'4332': {'pk': 4332, 'version': '24'}}, '3.15.1': {'4463': {'pk': 4463, 'version': '25'}}, '3.15.3': {'4582': {'pk': 4582, 'version': '26'}}, '3.16': {'4763': {'pk': 4763, 'version': '27'}}, '3.18': {'5205': {'pk': 5205, 'version': '28'}, '5320': {'pk': 5320, 'version': '29'}}, '3.20': {'5764': {'pk': 5764, 'version': '30'}}, '3.21.2': {'5980': {'pk': 5980, 'version': '31'}}, '3.22': {'6209': {'pk': 6209, 'version': '32'}}, '3.24': {'6684': {'pk': 6684, 'version': '33'}, '6867': {'pk': 6867, 'version': '34'}}, '3.26': {'7416': {'pk': 7416, 'version': '35'}, '7470': {'pk': 7470, 'version': '36'}}, '3.28': {'8110': {'pk': 8110, 'version': '37'}, '8121': {'pk': 8121, 'version': '38'}}, '3.30': {'8381': {'pk': 8381, 'version': '39'}}, '3.32': {'9396': {'pk': 9396, 'version': '40'}, '10201': {'pk': 10201, 'version': '41'}}, '3.34': {'11903': {'pk': 11903, 'version': '42'}, '13363': {'pk': 13363, 'version': '43'}}, '3.36': {'14372': {'pk': 14372, 'version': '44'}}, '3.38': {'19636': {'pk': 19636, 'version': '45'}, '21648': {'pk': 21648, 'version': '46'}}, '40.0': {'22834': {'pk': 22834, 'version': '47'}}, '40': {'24975': {'pk': 24975, 'version': '48'}}, '41': {'26942': {'pk': 26942, 'version': '49'}, '28466': {'pk': 28466, 'version': '50'}}, '42': {'29906': {'pk': 29906, 'version': '51'}}, '43': {'34582': {'pk': 34582, 'version': '52'}}, '44': {'40512': {'pk': 40512, 'version': '53'}}, '45': {'43704': {'pk': 43704, 'version': '54'}, '44582': {'pk': 44582, 'version': '55'}, '48949': {'pk': 48949, 'version': '56'}}, '46': {'51672': {'pk': 51672, 'version': '57'}, '52131': {'pk': 52131, 'version': '58'}}}">
```

Don't know if this is useful, though, parsing HTML in bash is a lost cause.

xynydev replied:
Actually, I was able to get that `data-svm` attribute with `xmllint` & `xpath` from a locally downloaded `.html` file.

`xmllint --html --xpath "/html/body//div[contains(@class, 'extension')]/@data-svm" ext6.html`

The output is still somewhat tough to parse, though.
```
ext6.html:40: HTML parser error : Tag header invalid
    <header class="gnome-header">
                                ^
 data-svm="{&quot;3.2&quot;: {&quot;pk&quot;: 284, &quot;version&quot;: 5}, &quot;3.2.0&quot;: {&quot;pk&quot;: 6, &quot;version&quot;: 1}, &quot;3.3.2&quot;: {&quot;pk&quot;: 33, &quot;version&quot;: 3}, &quot;3.2.1&quot;: {&quot;pk&quot;: 52, &quot;version&quot;: 4}, &quot;3.2.3&quot;: {&quot;pk&quot;: 284, &quot;version&quot;: 5}, &quot;3.3.4&quot;: {&quot;pk&quot;: 640, &quot;version&quot;: 7}, &quot;3.3.90&quot;: {&quot;pk&quot;: 831, &quot;version&quot;: 8}, &quot;3.4&quot;: {&quot;pk&quot;: 1008, &quot;version&quot;: 9}, &quot;3.5.5&quot;: {&quot;pk&quot;: 1690, &quot;version&quot;: 10}, &quot;3.5.90&quot;: {&quot;pk&quot;: 1736, &quot;version&quot;: 11}, &quot;3.5.91&quot;: {&quot;pk&quot;: 1807, &quot;version&quot;: 12}, &quot;3.6&quot;: {&quot;pk&quot;: 2213, &quot;version&quot;: 14}, &quot;3.7.5&quot;: {&quot;pk&quot;: 2489, &quot;version&quot;: 16}, &quot;3.7.90&quot;: {&quot;pk&quot;: 2560, &quot;version&quot;: 17}, &quot;3.7.92&quot;: {&quot;pk&quot;: 2640, &quot;version&quot;: 18}, &quot;3.8&quot;: {&quot;pk&quot;: 3216, &quot;version&quot;: 21}, &quot;3.10&quot;: {&quot;pk&quot;: 3247, &quot;version&quot;: 22}, &quot;3.12&quot;: {&quot;pk&quot;: 3739, &quot;version&quot;: 23}, &quot;3.14&quot;: {&quot;pk&quot;: 4332, &quot;version&quot;: 24}, &quot;3.15.1&quot;: {&quot;pk&quot;: 4463, &quot;version&quot;: 25}, &quot;3.15.3&quot;: {&quot;pk&quot;: 4582, &quot;version&quot;: 26}, &quot;3.16&quot;: {&quot;pk&quot;: 4763, &quot;version&quot;: 27}, &quot;3.18&quot;: {&quot;pk&quot;: 5320, &quot;version&quot;: 29}, &quot;3.20&quot;: {&quot;pk&quot;: 5764, &quot;version&quot;: 30}, &quot;3.21.2&quot;: {&quot;pk&quot;: 5980, &quot;version&quot;: 31}, &quot;3.22&quot;: {&quot;pk&quot;: 6209, &quot;version&quot;: 32}, &quot;3.24&quot;: {&quot;pk&quot;: 6867, &quot;version&quot;: 34}, &quot;3.26&quot;: {&quot;pk&quot;: 7470, &quot;version&quot;: 36}, &quot;3.28&quot;: {&quot;pk&quot;: 8121, &quot;version&quot;: 38}, &quot;3.30&quot;: {&quot;pk&quot;: 8381, &quot;version&quot;: 39}, &quot;3.32&quot;: {&quot;pk&quot;: 10201, &quot;version&quot;: 41}, &quot;3.34&quot;: {&quot;pk&quot;: 13363, &quot;version&quot;: 43}, &quot;3.36&quot;: {&quot;pk&quot;: 14372, &quot;version&quot;: 44}, &quot;3.38&quot;: {&quot;pk&quot;: 21648, &quot;version&quot;: 46}, &quot;40.0&quot;: {&quot;pk&quot;: 22834, &quot;version&quot;: 47}, &quot;40&quot;: {&quot;pk&quot;: 24975, &quot;version&quot;: 48}, &quot;41&quot;: {&quot;pk&quot;: 28466, &quot;version&quot;: 50}, &quot;42&quot;: {&quot;pk&quot;: 29906, &quot;version&quot;: 51}, &quot;43&quot;: {&quot;pk&quot;: 34582, &quot;version&quot;: 52}, &quot;44&quot;: {&quot;pk&quot;: 40512, &quot;version&quot;: 53}, &quot;45&quot;: {&quot;pk&quot;: 48949, &quot;version&quot;: 56}, &quot;46&quot;: {&quot;pk&quot;: 52131, &quot;version&quot;: 58}}"
```

Your approach is lean and mean and not prone to breakage, so I'm not sure whether this additional automation would be worth it.

fiftydinar replied:
It would be indeed great If we could implement this cleanly.

Is there something like `yq` but for `html` files?

Is `xmllint` what I basically ask?

xynydev replied:
> Is there something like yq but for html files?

`xq`

> Is xmllint what I basically ask?

It's an xml formatter and validator that apparently is both installed by default and can be used to parse HTML.

fiftydinar replied:
I will resolve this for now, as it's too difficult to implement.

fiftydinar committed:
feat: Implement compatibility check between extension & Gnome

If non-compatible version is detected from metadata.json, build will fail.

fiftydinar committed:
docs: Make advantages compared to RPMs clearer

fiftydinar committed:
feat: Fail the module if some crucial metadata.json info is missing

fiftydinar commented:
I found out that GSConnect doesn't have `settings-schema` key in `metadata.json` for some reason, so the module would fail to copy gschema xml.

The workaround would be to manually supply the key, to make extension work.
I will have to think on how the recipe YML would look with this feature If it's desired to implement.

Real solution is for developer to implement it.

I added checks for `shell-version`, `uuid` & `settings-schema` key in metadata.json, so if any of those are not present in the extension, build will fail.

EDIT: This issue is solved, I got rid of `settings-schema` dependency. I can just copy .gschema.xml without knowing the source name!
So basically all extensions are supported now.

fiftydinar committed:
docs: Add note in Known Issues that safe-check is implemented

fiftydinar committed:
fix: Don't repeat "Gnome version" log multiple times

fiftydinar committed:
docs: Rephrase Gnome compatibility sentence better

fiftydinar committed:
feat: Add full extension name for most error logs

fiftydinar committed:
fix: Get rid of SCHEMA_ID dependency!

gschema.xml can be copied without knowing the original name.

fiftydinar committed:
docs: "Almost all" extensions compatibility is true now.

fiftydinar committed:
style: Add more info about which extension is installed in logs

fiftydinar committed:
fix: Get rid of `gettext-domain` key, `install` key is already sufficient

fiftydinar committed:
feat: Copy icons if they exist & add better checks for gschema

fiftydinar committed:
style: Make icons log more readable

fiftydinar committed:
fix: Revert icons installation

It's more reliable to have them in main extension folder instead

They appear to work.

fiftydinar committed:
docs: Update "Usage" with more accurate steps on how to get the recipe input

fiftydinar commented:
I think that this is ready for review.

I currently didn't find any other way to improve the module, while not making the maintenance much harder.

xynydev commented:
Looks great overall! I'll review with a couple small things.

xynydev commented on modules/gnome-extensions/README.md:
```diff
@@ -0,0 +1,30 @@
+The only maintenance is to bump the extension version when new Fedora/Gnome releases (around every 6 months).
+
+What does this module do?
+- It parses the gettext-domain that you inputted, along with the extension version
```
This line is probably opaque and meaningless to most readers. What is a gettext-domain? Even a Google search only leads to some results that don't really help with understanding the concept, if they are even related.

Should either explain more, or less. But the middleground with giving too much information but not what is needed to understand it is not helpful as the first element in this list.

fiftydinar replied:
I got rid of the technical terms & only explained it basically, but still technically correct, so all users can understand.

xynydev commented on modules/gnome-extensions/README.md:
```diff
@@ -0,0 +1,30 @@
+Almost all Gnome extensions are compatible for installation.  
+Only rare intervention that might be needed is for extensions which require some additional system dependencies, like Pano.   
+
+Thanks to https://extensions.gnome.org which provides end-releases of extensions as zips, it is very easy to maintain this module configuration.  
```
```suggestion
Thanks to https://extensions.gnome.org which provides releases of extensions as zips, it is very easy to maintain this module configuration.  
```

xynydev commented on modules/gnome-extensions/README.md:
```diff
@@ -0,0 +1,30 @@
+This module is universally compatible with all distributions which ship Gnome, as it's not tied to specific distribution packaging format for installing extensions.  
+
+Almost all Gnome extensions are compatible for installation.  
+Only rare intervention that might be needed is for extensions which require some additional system dependencies, like Pano.   
```
> like Pano.

Should add a link to "Pano" and instructions for what to do when an extension needs this "rare intervention".

fiftydinar replied:
I got rid of "Pano", in case it gets depreciates, so there is no need to edit the README for that.

I added the explanation in Usage, as I find it fits better there.

fiftydinar committed:
docs: Use "releases" term instead of "end-releases"

Co-authored-by: xyny <60004820+xynydev@users.noreply.github.com>

xynydev commented on modules/gnome-extensions/README.md:
```diff
@@ -0,0 +1,30 @@
+
+# Usage
+
+To use this module, you need to input gettext-domain of the extension without @ symbol + the version of the extension in `.v%VERSION%` format in module recipe.  
```
The string `%VERSION%` should not be used to represent the extension version unless it is actually what the user should input in the recipe to get some sort of dynamic version. This is because the `rpm-ostree` module uses `%OS_VERSION%` as a shorthand for the current Fedora version in repo URLs, and using the same syntax to mean different things might be confusing.

fiftydinar replied:
I think that this part of the section can be removed with your latest doc changes, as everything necessary is explained in install instructions.

fiftydinar replied:
Resolved

xynydev commented on modules/gnome-extensions/README.md:
```diff
@@ -0,0 +1,30 @@
+But for some extensions, `.v%VERSION%` is parsed as the version of package in incremental numbering, instead of extension version in https://extensions.gnome.org URL.  
+So to be sure that you got the correct module input, follow the steps below.
+
+How to gather correct module input:  
```
```suggestion
How to install extensions using the module:  
```

~~Maybe this would be a good way to replace the "gettext domain" term?~~

Edit: changed section title to reflect the content in another suggestion.

xynydev replied:
Or actually, the extensions already have IDs, so IDK:

`https://extensions.gnome.org/extension/2236/night-theme-switcher/`
-> id is `2236`

xynydev replied:
Either way, `module input` and `gettext domain` should both be replaced with a better term if we can come up with one.

xynydev commented on modules/gnome-extensions/README.md:
```diff
@@ -0,0 +1,30 @@
+1. Go to https://extensions.gnome.org
+2. Search for the extension that you want
+3. Select the matching Gnome shell version & extension version that you want to download
+4. When extension is downloaded, you get the info when you omit `.shell-extension.zip` suffix from the extension zip file-name
```
```suggestion
1. Go to https://extensions.gnome.org
2. Search for the extension that you want to install and open its extension page
3. Select the correct GNOME shell version & extension version from the dropdown
   - The command `gnome-shell --version` can be used to get the GNOME version of a running system.
4. When the download dialog for the extension comes up, copy everything but the `.shell-extension.zip` suffix from the filename into the `install:` array in this module's configuration.
```

fiftydinar committed:
docs: Add clearer Usage instructions

Co-authored-by: xyny <60004820+xynydev@users.noreply.github.com>

fiftydinar committed:
docs: Change title to be clearer & to avoid `gettext-domain` & `module-input` terms

Co-authored-by: xyny <60004820+xynydev@users.noreply.github.com>

fiftydinar committed:
style: Make "DONE" in logs symmetric & get rid of `gettext-domain` term

fiftydinar committed:
docs: Get rid of redundant Usage instructions

fiftydinar committed:
docs: Add explanation on how to install additional system dependencies if needed for extension

fiftydinar committed:
docs: Get rid of unnecessary comments in module.yml

fiftydinar committed:
docs: Fix grammar in logs for non-specified module input

xynydev commented on modules/gnome-extensions/README.md:
```diff
@@ -25,3 +24,7 @@ How to install extensions using the module:
+
+Rarely, some extensions need additional system dependencies in order to function.  
+Those extensions usually note that case inside the extension description webpage.  
+The solution is to install the required dependencies by using `rpm-ostree` module before `gnome-extensions` module is ran.
```
```suggestion
An extension might need additional system dependencies in order to function.  
In that case, you should install the required dependencies before the `gnome-extensions` module is ran.
Information about the required dependencies (if any) are usually on the extension's page.  
```

fiftydinar committed:
docs: Make extensions with additional dependencies section even better

Co-authored-by: xyny <60004820+xynydev@users.noreply.github.com>
